### ISSUE TITLE
Implementation of Deserialize and Decode is Not General Enough

### ISSUE THREAD

**Issue by hqdncw** [matteopolak/axum-codec#4]
```rust
use axum_codec::{Codec, ContentType};

#[derive(Debug, PartialEq)]
#[axum_codec::apply(decode, encode)]
struct MyStruct<'a> {
    message: &'a str,
}

fn main() {
    Codec::<MyStruct>::from_bytes(&[], ContentType::Bitcode);  // throws a compile error
}
```

```bash
Diagnostics:
1. implementation of `Deserialize` is not general enough
   `MyStruct<'_>` must implement `Deserialize<'0>`, for any lifetime `'0`...
   ...but `MyStruct<'_>` actually implements `Deserialize<'1>`, for some specific lifetime `'1`
2. implementation of `Decode` is not general enough
   `MyStruct<'_>` must implement `Decode<'0>`, for any lifetime `'0`...
   ...but `MyStruct<'_>` actually implements `Decode<'1>`, for some specific lifetime `'1`
   ```

**Comment by matteopolak** [comment]
`ciborium`'s lack of support for borrowing from the input is currently blocking the ability to use `Deserialize` instead of `DeserializeOwned`. Since this is a pretty common use-case I think I'll just remove `cbor` support for now.

Related: https://github.com/enarx/ciborium/issues/36

**Comment by matteopolak** [comment]
Your code sample now compiles. Note that decoding into a `&'a str` will panic when using `toml` and `msgpack` in some scenarios. I would recommend using `Cow<'a, str>` (see the mentioned PR for more information).

Thanks for the issue :)

**Comment by hqdncw** [comment]
Woo-hoo! That was so quick, @matteopolak! Thanks a lot for the update and the tip about using `Cow<'a, str>`.

**Comment by hqdncw** [comment]
```toml
# Cargo.toml

[dependencies}
axum-codec = { version = "0.0.15" }
```

```rust
use axum::Router;
use axum_codec::Codec;

#[axum_codec::apply(decode, encode)]
pub struct MyRequest<'a> {
    message: &'a str,
}

pub async fn ping(Codec(_): Codec<MyRequest<'_>>) -> () {}

#[tokio::main]
async fn main() {
    let app = Router::new()
        .route("/", axum_codec::routing::get(ping).into()); // throws a compile error

    let listener = tokio::net::TcpListener::bind("0.0.0.0:3000").await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
```

```bash
Diagnostics:
1. implementation of `CodecDecode` is not general enough
   `MyRequest<'_>` must implement `CodecDecode<'de>`
   ...but it actually implements `CodecDecode<'0>`, for some specific lifetime `'0`
 ```

**Comment by matteopolak** [comment]
I'll look into this soon, just finishing up my finals.

I think the easiest way to resolve this is to have a `BorrowCodec` that stores the body (similar approach to axum's `JsonDeserialize`). Destructuring it will be a bit awkward so I'll look into a few designs and get back to you.

**Comment by hqdncw** [comment]
There's also [ouroboros](https://github.com/someguynamedjosh/ouroboros). It's a library that allow you to bundle owned data with a borrowing type, which might be applicable here, but I'm not sure how well it would fit this specific use case.

**Comment by matteopolak** [comment]
Added an example (see linked PR)

```rust
#[axum_codec::apply(encode, decode)]
struct BorrowGreeting<'d> {
	#[serde(borrow)]
	message: Cow<'d, str>,
}

async fn borrow_greet(
	accept: Accept,
	greeting: BorrowCodec<BorrowGreeting<'_>>,
) -> Result<impl IntoCodecResponse, CodecRejection> {
	let greeting = greeting.decode()?;
	let is_zero = matches!(greeting.message, Cow::Borrowed(..));

	Ok(Codec(Greeting {
		message: if is_zero {
			"Borrowing from input"
		} else {
			"Not borrowing from input (for JSON, probably using an escaped character)"
		}
		.to_string(),
	}))
}
```

I added some comments in the PR, but if this looks like a good start to you I'll merge this in. In the future I'll look for a way to keep it all as a self-contained extractor (without needing to `greeting.decode()?`

**Comment by matteopolak** [comment]
I decided to rewrite it so it closely follows the usage of `Codec`:

```rust
#[axum_codec::apply(encode, decode)]
struct BorrowGreeting<'d> {
	#[serde(borrow)]
	message: Cow<'d, str>,
}

async fn borrow_greet(
	greeting: BorrowCodec<BorrowGreeting<'_>>,
) -> impl IntoCodecResponse {
	println!("{:?}", greeting.message);
}
```

I added some miri tests. I'll need to make sure the implementation is sound in regards to `Pin<Bytes>`, but there shouldn't be any other major changes.

**Comment by matteopolak** [comment]
Released in `v0.0.16`. Please reopen if there's anything I missed
